nestjs/src/schedule/schedule.service.ts

```diff
@@ -2,49 +2,65 @@ import { Injectable, Logger } from '@nestjs/common';
 import { CoursesService } from 'src/courses/courses.service';
 import { CourseTasksService } from 'src/courses';
 import { Student } from '@entities/student';
+import { Mentor } from '@entities/mentor';
+import { unionBy } from 'lodash';
+import { StudentsService } from 'src/courses/students';
+import { MentorsService } from 'src/courses/mentors';
 
 @Injectable()
 export class ScheduleService {
   private readonly logger = new Logger('schedule');
-  constructor(private courseService: CoursesService, private courseTasksService: CourseTasksService) {}
+  constructor(
+    private courseService: CoursesService,
+    private courseTasksService: CourseTasksService,
+    private studentsService: StudentsService,
+    private mentorsService: MentorsService,
+  ) {}
 
-  public async getChangedStudentsCourses() {
-    const activeCourses = await this.courseService.getActiveCourses(['students']);
+  public async getChangedCoursesRecipients() {
+    const activeCourses = await this.courseService.getActiveCourses();
 
     const updatedCourseSchedule = [];
     for (const course of activeCourses) {
       const tasks = await this.courseTasksService.getUpdatedTasks(course.id, 1);
 
       const updatedTasks = tasks.length > 0;
 
+      const [students, mentors] = await Promise.all([
+        this.studentsService.getActiveByCourseId(course.id),
+        this.mentorsService.getActiveByCourseId(course.id),
+      ]);
+
       this.logger.log({ message: `course: ${course.id} has ${tasks.length} tasks updated` });
 
       if (updatedTasks) {
         updatedCourseSchedule.push({
           alias: course.alias,
           name: course.name,
-          students: course.students.filter(student => !student.isExpelled),
+          students,
+          mentors,
         });
       }
     }
 
-    return this.mapChangedCourseToStudents(updatedCourseSchedule);
+    return this.mapChangedCoursesToRecipients(updatedCourseSchedule);
   }
 
-  private mapChangedCourseToStudents(courses: { alias: string; students: Student[] }[]) {
-    const studentMap = new Map<number, { alias: string }[]>();
+  private mapChangedCoursesToRecipients(courses: { alias: string; students: Student[]; mentors: Mentor[] }[]) {
+    const recipientMap = new Map<number, { alias: string }[]>();
 
     for (const course of courses) {
-      const { students, ...rest } = course;
+      const { students, mentors, ...rest } = course;
+      const users = unionBy<{ userId: number }>(students, mentors, 'userId');
 
-      for (const student of students) {
-        const studentCourses = studentMap.get(student.userId) ?? [];
+      for (const user of users) {
+        const userCourses = recipientMap.get(user.userId) ?? [];
```

Comment:
I wondering can it be done faster by sql-request?

Reply:
what exactly? could you propose?

Reply:
seems i understood what you mean. will update

Reply:
I mean instead of all this operations with receiving students, mentors and users separately and putting it together with JS make something like this:
```sql
SELECT
  u.id,
  array_agg(coalesce(DISTINCT c."alias", '{}'::text)) as "courses"
FROM "user" u
LEFT JOIN student s ON u.id = s."userId"
LEFT JOIN mentor  m ON u.id = m."userId"
LEFT JOIN course  c ON c.id = s."courseId" OR c.id = m."courseId"
WHERE
  s."courseId" = $1 AND s."isExpelled" = false OR
  m."courseId" = $1 AND m."isExpelled" = false
GROUP BY u.id
```
And then just put it to the Map:
```javascript
const recipientMap = new Map<number, { alias: string }[]>(users.map(({ id, courses }) => [id, courses]);
```

Reply:
updated